devtools/shared/invoke-planner.js

```diff
@@ -0,0 +1,92 @@
+/**
+ * @license
+ * Copyright (c) 2018 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+let planner;
+
+function getStrategies() {
+  if (window.Arcs == undefined || window.arc == undefined)
+    return null;
+  planner = new Arcs.Planner();
+
+  planner.init(arc);
+  return planner.strategizer._strategies.map(a => a.constructor.name);
+  
+}
+
+async function invokePlanner(stageName, inputRecipe) {
+  if (typeof window !== 'object')
+    return;
+
+  let strategy = planner.strategizer._strategies.find(a => a.constructor.name == stageName);
+
+  if (!strategy)
+    return {error: 'could not find strategy'}; 
+  
+  let manifest = await Arcs.Manifest.parse(inputRecipe, {loader: arc._loader, fileName: 'manifest.manifest'});
+  let recipe = manifest.recipes[0];
+  recipe.normalize();
+
+  let strategizer = {generated: [{result: recipe, score: 1}], generation: 0};
+  let {results} = await strategy.generate(strategizer);
+  if (results.length) {
+    for (let result of results) {
+      result.hash = await result.hash;
+      result.derivation = undefined;
+      let recipe = result.result;
+      result.result = recipe.toString({showUnresolved: true});
+      let errors = new Map();
+      recipe.normalize({errors});
+      result.errors = [...errors.keys()].map(thing => ({id: thing.id, error: errors.get(thing)}));
+      result.normalized = recipe.toString();
+    }
+  }
+  return results;
+}
+
+function getStrategiesAndSend() {
+  console.log('gSAS');
+  let strategies = getStrategies();
+  if (strategies == null) {
+    setTimeout(getStrategiesAndSend, 500);
+    return;
+  }
+  console.log(strategies);
+  sendMessage({op: 'strategies', strategies});
+}
+
+async function handlePlannerRequest(message) {
+  console.log(message);
+  switch (message.op) {
+    case 'fetch-strategies':
+      getStrategiesAndSend();
+      break;
+    case 'invoke':
+      let results = await invokePlanner(message.data.strategy, message.data.recipe);
+      sendMessage({op: 'planner-results', results});
+  } 
+}
```

Comment:
this adds more ad-hoc marshalling/unmarshalling. Should I look at building a more consistent end-to-end communication channel?

Reply:
Communication in the arcs->devtools direction was supposed to happen via the platform/devtools-channel-{web,node}, accessible by runtime/debug/devtools-channel-provider. You would probably run into problems with using it here, as platform switching is not done, and it's not bundled with the extension etc.

I have a hunch this file should live in runtime/debug directory and be built with runtime proper. This would make it possible to import whatever components you need and not rely on window.Arcs. You could also use the existing infrastructure for the Arcs->DevTools communication. This in turn could make it work out of the box with Arcs in NodeJS as well.

The problem of delivering messages from DevTools to Arcs only once the page is booted and Arcs is initialized remains, but we can surely solve it in the abstract way in the message passing channel.

WDYT?